README.md
# Coherence

[![Build Status](https://travis-ci.org/smpallen99/coherence.svg?branch=master)](https://travis-ci.org/smpallen99/coherence) [![Hex Version][hex-img]][hex] [![License][license-img]][license]

[hex-img]: https://img.shields.io/hexpm/v/coherence.svg
[hex]: https://hex.pm/packages/coherence
[license-img]: http://img.shields.io/badge/license-MIT-brightgreen.svg
[license]: http://opensource.org/licenses/MIT

Checkout the [Coherence Demo Project](https://github.com/smpallen99/coherence_demo) to see an example project using Coherence.

Coherence is a full featured, configurable authentication system for Phoenix, with the following modules:

* [Database Authenticatable](#authenticatable): handles hashing and storing an encrypted password in the database.
* [Invitable](#invitable): sends invites to new users with a sign-up link, allowing the user to create their account with their own password.
* [Registerable](#registerable): allows anonymous users to register a users email address and password.
* [Confirmable](#confirmable): new accounts require clicking a link in a confirmation email.
* [Recoverable](#recoverable): provides a link to generate a password reset link with token expiry.
* [Trackable](#trackable): saves login statistics like login counts, timestamps, and IP address for each user.
* [Lockable](#lockable): locks an account when a specified number of failed sign-in attempts has been exceeded.
* [Unlockable With Token](#unlockable-with-token): provides a link to send yourself an unlock email.
* [Rememberable](#remember-me): provides persistent login with 'Remember me?' check box on login page.

Coherence provides flexibility by adding namespaced templates and views for only the options specified by the `mix coh.install` command. This boiler plate code is added to your `lib/my_project/web/templates/coherence` and `lib/my_project/web/views/coherence` directories.

Once the boilerplate has been generated, you are free to customize the source as required.

As well, a `lib/my_project/web/coherence_web.ex` is added. Migrations are also generated to add the required database fields.

See the [Docs](https://hexdocs.pm/coherence/Coherence.html) and [Wiki](https://github.com/smpallen99/coherence/wiki) for more information.

## Installation

  1. Add coherence to your list of dependencies in `mix.exs`:

      ```elixir
      def deps do
        [{:coherence, "~> 0.8"}]
      end
      ```

  2. Ensure coherence is started before your application:

      ```elixir
      def application do
        extra_applications: [..., :coherence]]
      end
      ```

## Upgrading

After upgrading a Coherence version, you should generate the boilerplate files. To assist this process, use the `--reinstall` option.

This option uses your project's existing coherence config and runs the the installer with the same options.

```shell
mix coh.install --reinstall
```

Run a `git diff` to review the updated files. If you had updated any of the boilerplate files, you may need to manually integrate the changes into the newly generated files.

Run `mix help coh.install` for more information.

## Phoenix & Phx Project Structure

Coherence supports projects created with the older `mix phoenix.new` and the newer `mix phx.new` commands. Separate versions of the mix tasks exist for each project structure.

For projects created with `mix phx.new`, use the following mix tasks:

* `coh.install`
* `coh.clean`

For projects created with `mix phx.new --umbrella`, ensure you are in the app directory and use the following options for the install:
* `cd apps/my_project`
* `coh.install --web-module MyProjectWeb --web-path ../my_project_web/lib/my_project_web`

## Getting Started

First, decide which modules you would like to use for your project. For the following example we are going to use a full install except for the confirmable option.

Run the installer

```bash
$ mix coh.install --full-invitable
```

This will:

* add the coherence configuration to the end of your `config/config.exs` file.
* add a new User model if one does not already exist
* add migration files
  * timestamp_add_coherence_to_user.exs if the User model already exists
  * timestamp_create_coherence_user.exs if the User model does not exist
  * timestamp_create_coherence_invitable.exs
* add view files lib/my_project/web/views/coherence/
* add template files to lib/my_project/web/templates/coherence
* add email files to lib/my_project/web/emails/coherence
* add lib/my_project/web/coherence_web.ex file
* add lib/my_project/web/coherence_messages.ex file

You should review your `config/config.exs` as there are a couple items you will need to customize like email address and mail api_key. If you don't edit the email_from value to something different than its default, emails may not be sent.

See [Installer](#installer) for more install options.

You will need to update a few files manually.

```elixir
# lib/my_project_web/router.ex

defmodule MyProjectWeb.Router do
  use MyProjectWeb, :router
  use Coherence.Router         # Add this

  pipeline :browser do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug Coherence.Authentication.Session  # Add this
  end

  # Add this block
  pipeline :protected do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug Coherence.Authentication.Session, protected: true
  end

  # Add this block
  scope "/" do
    pipe_through :browser
    coherence_routes()
  end

  # Add this block
  scope "/" do
    pipe_through :protected
    coherence_routes :protected
  end

  scope "/", MyProjectWeb do
    pipe_through :browser

    get "/", PageController, :index
    # add public resources below
  end

  scope "/", MyProjectWeb do
    pipe_through :protected

    # add protected resources below
    resources "/privates", MyProjectWeb.PrivateController
  end
end
```
**Important**: Note the name-spacing above. Unless you generate coherence controllers, ensure that the scopes, `scope "/" do`, do not include your projects' scope here. If so, the coherence routes will not work!

If the installer created a user schema (one did not already exist), there is nothing you need to do with that generated file. Otherwise, update your existing schema (assuming its `Accounts.User` like this:

```elixir
# lib/my_project/accounts/user.ex

defmodule MyProject.Accounts.User do
  use Ecto.Schema
  use Coherence.Schema                                    # Add this

  schema "users" do
    field :name, :string
    field :email, :string
    coherence_schema()                                    # Add this

    timestamps()
  end

  def changeset(model, params \\ %{}) do
    model
    |> cast(params, [:name, :email] ++ coherence_fields)  # Add this
    |> validate_required([:name, :email])
    |> validate_format(:email, ~r/@/)
    |> validate_coherence(params)                         # Add this
  end

  def changeset(model, params, :password) do
    model
    |> cast(params, ~w(password password_confirmation reset_password_token reset_password_sent_at))
    |> validate_coherence_password_reset(params)
  end
end
```

An alternative approach is add the authentication plugs to individual controllers that require authentication. You will want to use this approach if you require authentication for a subset of actions in a controller.

For example, lets say you want to show a list of products for everyone visiting the site, but only want authenticated users to be able to create, update, and delete products. You could do the following:

Ensure the following is in your `lib/my_project_web/router.ex` file:

```elixir
scope "/", MyProjectWeb do
  pipe_through :browser
  resources "/products", ProductController
end
```

In your product controller add the following:

```elixir
defmodule MyProjectWeb.ProductController do
  use MyProjectWeb, :controller

  plug Coherence.Authentication.Session, [protected: true] when action != :index

  # ...
end
```

## Default Configuration

```elixir
{:require_current_password, true}, # Current password is required when updating new password.
{:reset_token_expire_days, 2},
{:confirmation_token_expire_days, 5},
{:allow_unconfirmed_access_for, 0},
{:max_failed_login_attempts, 5},
{:unlock_timeout_minutes, 20},
{:unlock_token_expire_minutes, 5},
{:rememberable_cookie_expire_hours, 2*24},
{:forwarded_invitation_fields, [:email, :name]}
{:allow_silent_password_recovery_for_unknown_user, false},
{:password_hashing_alg, Comeonin.Bcrypt}
```

You can override this default configs. For example: you can add the following codes inside `config/config.exs`

```elixir
config :coherence,
  require_current_password: false,
  max_failed_login_attempts: 3
```

## Custom registration and sessions routes

Coherence supports custom routes for registration and login. These configurations can be set globally or scoped.

Which routes can be custom?

```
%{
  registrations_new:  "/registrations/new",
  registrations:      "/registrations",
  passwords:          "/passwords",
  confirmations:      "/confirmations",
  unlocks:            "/unlocks",
  invitations:        "/invitations",
  invitations_create: "/invitations/create",
  invitations_resend: "/invitations/:id/resend",
  sessions:           "/sessions",
  registrations_edit: "/registrations/edit"
}
```

To set them globally add the following to you configuration:



```elixir
config :coherence,
  default_routes: %{
    registrations_edit: "/accounts/edit", ...},
```

To set them scoped for each mode (protected, public, etc..):

```elixir
scope "/" do
  pipe_through :protected
  coherence_routes :protected, [
    custom_routes: %{registratons_edit: "/accounts/edit", ...}
  ]
end
```

## Phoenix Channel Authentication

Coherence supports channel authentication using `Phoenix.Token`. To enable channel authentication do the following:

Add the following option to coherence configuration.

```elixir
config :coherence,
  user_token: true
```

Update your socket module:

```elixir
defmodule MyProjectWeb.UserSocket do
  use Phoenix.Socket

  def connect(%{"token" => token}, socket) do
    case Coherence.verify_user_token(socket, token, &assign/3) do
      {:error, _} -> :error
      {:ok, socket} -> {:ok, socket}
    end
  end
end
```

## Localization with Gettext

Coherence supports `Gettext` for all User facing messages. Your project's `Gettext` module is used by default.

All Coherence messages use `dgettext` with the `"coherence"` domain. This means that after you run `mix gettext.extract`, you will see `coherence.pot` files generated. Running `mix gettext.merge priv/gettext` will generate the corresponding `coherence.po` files.

Coherence does not come with pre-translated `.po` files. We figure that you will want to tweak the Coherence language to suite your application.

All files added to your project with the Coherence generators include a `import MyProjectWeb.Gettext` with the `"coherence"` domain.

The other messages that Coherence uses internal are pulled from the `my_project_web/coherence_messages.ex` file that is generated with the `coh.install` mix task. You can edit this file and customize it as required.

To assist in upgrades to future releases of Coherence, the `MyProject.Coherence.Messages` module uses the `Coherence.Messages` behaviour which defines each message function. So, if you miss a message after upgrading, you will see a message during compile time.

## Option Overview

### Authenticatable

Handles hashing and storing an encrypted password in the database.

Provides `/sessions/new` and `/sessions/delete` routes for logging in and out with
the appropriate templates and view.

The following columns are added the `<timestamp>_add_coherence_to_user.exs` migration:

* :password_hash, :string - the encrypted password

### Invitable

Handles sending invites to new users with a sign-up link, allowing the user to create their account with their own password.

Provides `/invitations/new` and `invitations/edit` routes for creating a new invitation and creating a new account from the invite email.

These routes can be configured to require login by using the `coherence_routes :private` macro in your router.exs file.

Invitation token timeout will be added in the future.

The following table is created by the generated `<timestamp>_create_coherence_invitable.exs` migration:

```elixir
create table(:invitations) do
  add :name, :string
  add :email, :string
  add :token, :string
end
```

### Registerable

Allows anonymous users to register a users email address and password.

Provides `/registrations/new`, `create`, `edit`, `update`, `show`, and `delete` routes for managing registrations.

Adds the following:

* `Register New Account` to the log-in page.
* `Link to account page on layout helpers`
* Show page with `edit` and `delete` links
* `edit` page

It is recommended that the :confirmable option is used with :registerable to
ensure a valid email address is captured.

### Confirmable

Requires a new account be confirmed. During registration, a confirmation token is generated and sent to the registering email. This link must be clicked before the user can sign-in.

Provides `edit` action for the `/confirmations` route.

The confirmation token expiry default of 5 days can be changed with the `:confirmation_token_expire_days` config entry.

### Recoverable

Allows users to reset their password using an expiring token send by email.

Provides `new`, `create`, `edit`, `update` actions for the `/passwords` route.

Adds a "Forgot your password?" link to the log-in form. When clicked, the user provides their email address and if found, sends a reset password instructions email with a reset link.

The expiry timeout can be changed with the `:reset_token_expire_days` config entry.

By default, providing an unknown email address will result in a form error. If you want to prevent that and display a confirmation message even if the email isn’t found, set the `:allow_silent_password_recovery_for_unknown_user` to `true`.

### Trackable

Saves login statistics like login counts, timestamps, and IP address for each user.

Adds the following database field to your User model with the generated migration:

```elixir
add :sign_in_count, :integer, default: 0  # how many times the user has logged in
add :current_sign_in_at, :datetime        # the current login timestamp
add :last_sign_in_at, :datetime           # the timestamp of the previous login
add :current_sign_in_ip, :string          # the current login IP adddress
add :last_sign_in_ip, :string             # the IP address of the previous login
```

### Lockable

Locks an account when a specified number of failed sign-in attempts has been exceeded.

The following defaults can be changed with the following config entries:

* `:unlock_timeout_minutes`
* `:max_failed_login_attempts`

Adds the following database field to your User model with the generated migration:

```elixir
add :failed_attempts, :integer, default: 0
add :unlock_token, :string
add :locked_at, :datetime
```

### Unlockable with Token

Provides a link to send yourself an unlock email. When the user clicks the link, the user is presented a form to enter their email address and password. If the token has not expired and the email and password are valid, a unlock email is sent to the user's email address with an expiring token.

The default expiry time can be changed with the `:unlock_token_expire_minutes` config entry.

### Remember Me

The `rememberable` option provides persistent login when the 'Remember Me?' box is checked during login.

With this feature, you will automatically be logged in from the same browser when your current login session dies using a configurable expiring persistent cookie.

For security, both a token and series number stored in the cookie on initial login. Each new creates a new token, but preserves the series number, providing protection against fraud. As well, both the token and series numbers are hashed before saving them to the database, providing protection if the database is compromised.

The following defaults can be changed with the following config entries:

* :rememberable_cookie_expire_hours (2\*24)
* :login_cookie                     ("coherence_login")

The following table is created by the generated `<timestamp>_create_coherence_rememberable.exs` migration:

```elixir
create table(:rememberables) do
  add :series_hash, :string
  add :token_hash, :string
  add :token_created_at, :datetime
  add :user_id, references(:users, on_delete: :delete_all)

  timestamps
end

create index(:rememberables, [:user_id])
create index(:rememberables, [:series_hash])
create index(:rememberables, [:token_hash])
create unique_index(:rememberables, [:user_id, :series_hash, :token_hash])
```

The `--rememberable` install option is not provided in any of the installer group options. You must provide the `--rememberable` option to install the migration and its support.

### User Active Field

The `--user-active-field` option can be given to the installer to support inactive users. When a user is set as inactive, they will not be able to login.

Using this option adds a `active` field to the user schema with a default of `true`.

## Mix Tasks

### Installer

The following examples illustrate various configuration scenarios for the install mix task:

```bash
# Install with only the `authenticatable` option
$ mix coh.install

# Install all the options except `confirmable` and `invitable`
$ mix coh.install --full

# Install all the options except `invitable`
$ mix coh.install --full-confirmable

# Install all the options except `confirmable`
$ mix coh.install --full-invitable

# Install the `full` options except `lockable` and `trackable`
$ mix coh.install --full --no-lockable --no-trackable
```

And some reinstall examples:

```bash
# Reinstall with defaults (--silent --no-migrations --no-config --confirm-once)
$ mix coh.install --reinstall

# Confirm to overwrite files, show instructions, and generate migrations
$ mix coh.install --reinstall --no-confirm-once --with-migrations
```

Run `$ mix help coh.install` for more information.

### Clean

The following examples illustrate how to remove the files created by the installer:

```bash
# Clean all the installed files
$ mix coh.clean --all

# Clean only the installed view and template files
$ mix coh.clean --views --templates

# Clean all but the models
$ mix coh.clean --all --no-models

# Prompt once to confirm the removal
$ mix coh.clean --all --confirm-once
```

After installation, if you later want to remove one more options, here are a couple examples:

```bash
# Clean one option

CHANGELOG.md
# Coherence Changelog

## 0.8.0 (2023-09-22)

* Enhancements
  * Support Elixir 1.14

* Bug Fixes
  * Resolve Dialyzer warnings, correct specs and typos


## 0.7.0 (2020-04-26)

* Enhancements
  * Support Elixir 1.10 and ecto 4


## 0.6.2 (2020-02-17)

* Bug Fixes
  * Clear user_data in Coherence.CredentialStore.Server where there are no entries for that user


## 0.6.1 (2019-11-04)

* Bug Fixes
  * Fix exception in Coherence.CredentialStore.Server.update_user_logins when invalid id


## 0.6.0 (2018-09-xx)

See these [0.5.x to 0.6.x upgrade instructions](https://gist.github.com/smpallen99/fc59423d521f7f6773a5fa9236512faf) to upgrade your existing apps.

* Enhancements
  * Moved Controller actions and helpers into a base modules inside __using__ macro to improve upgrades with custom controllers
  * Add configurable salt and max_age for tokens
  * Add default opts for token functions
  * Add support for application configurable confirmable
  * Support code formatter formatter
  * Add logging to the mailer template
  * Updated to Comeonin 4.0
  * Add support for configurable password hashing algorithms
  * Speed up tests by configuring the Bcrypt algorithm
  * Format may generator templates to align with code formatter

* Bug Fixes
  * Fixed detection of remember me checkbox on session new page
  * Fixed compiled gettext in view helpers
  * Fix new session screen issue on newly generated project #390
  * Add back the gettext fix in ViewHelpers

* Deprecations
  * Removed the coherence.make_templates task since its no longer needed with the new controller design

* Hard Deprecations
  * Removed support for erlang < 20.0
  * Removed support for pre Phoenix 1.3 project structure
    * removed `coherence.install` and `coherence.gen.controllers` mix tasks

## 0.5.2 (2018-09-03)

* Enhancements
  * Added default permitted attributes to help with upgrades #371

## 0.5.1 (2018-08-28)

* Enhancements
  * Add new `coh.gen.controllers` generator
  * `coh.clean` task now supports phx and phoenix projects
  * Renamed `Coherence.ControllerHelpers` module to `Coherence.Controller`
  * Update timex and timex_ecto dependency versions. closes #333
  * Added new update_user_login/1 function
  * Added new `allowed attributes` feature to protect against allowing accesses to internal schema fields.
  * Update uuid dependency version
  * Allow silent password recovery for unknown users

* Bug Fixes
  * Fixed issues with generated controllers
  * Fixed issues with clean tasks
  * Fixed localization of view helpers. closes #322 and closes #327
  * Fixed incorrect logic for custom password changeset. closes #351
  * Fixed config return type for email_reply_to. closes #227
  * Fixed :peer retrieval issue with new Plug.Conn
  * Fixed issue attempting to immediately log in after registering an account
  * Fix in password controller to support subdomains

* Backward incompatible changes
  * `--controllers` option is not supported for the install tasks
  * removed the `coherence.clean` task
  * Need to update `seeds.exs` if your using confirmable option

## 0.5.0 (2017-08-03)

* Enhancements
  * Support the released Version of Phoenix 1.3 project structure
  * Project name spaced generated view modules
  * GenServer based Session Store
    * Single user model stored for multiple logins of the same user
    * Support to update user model for all logged in sessions
  * add --web-module installer option
  * Generate Invitation, Rememberable, and Trackable schemas
  * Use app's layout by default. Use --layout option to generate and use a specific layout for coherence. #186
  * Remove login callback and replace with Phoenix.Controller.redirect. #254
  # Add support for active field in user schema. #201

* Bug Fixes
  * Fixed incorrect reference to CoherenceDemo in Config.mailer?
  * Proper support for binary_ids

* Backward incompatible changes
  * Does not support Phoenix 1.3.0-rc versions with lib/my_project/web project structure
  * Previous generated controllers and views must be updated

## 0.4.0 (2017-07-03)

See these `0.3.1` to `0.4.0` [upgrade instructions](https://gist.github.com/smpallen99/92d826b6f523edd83b4a358cf05e97e7) to bring your existing app up to speed.

* Enhancements
  * Localization with Gettext
  * Support the new `Phoenix 1.3` project structure created with `mix phx.new`. Use `mix coh.install`
  * Support for legacy projects with the `mix coherence.install` task
  * Callback to redirect invitation resend
  * Add use_binary_id config
  * Use the binary_id generators setting in your project
  * Remove most of the Credo warnings
  * Support user tokens for channel authentication
  * Removed compiler warnings for Elixir 1.4
  * Change datetime to utc_datetime
  * Added new trackable-table option to reduce noise in user schema
  * Track login and logout in trackable-table
  * Added specs and resolved many dialyzer warnings
  * Added asyc_rememberable? support for near concurrent Ajax requests
  * Don't increment rememberable sequence number for Ajax requests
  * Added support for custom layouts
  * Added configurable require current password to change passwords
  * Added support for fast switch user (no password required) in dev environment only

* Bug Fixes
  * Fix session controller when lockable is false
  * Make remember me clickable
  * Fixed layout for unauthenticated protected routes
  * Fixed race conditions in rememberable callback

* Backwards incompatible changes
  * coherence config requires `:messages_backend` field to be set (MyProject.Coherence.Messages)
  * coherence_messages.ex file must be generated and included in our app. Done by the installer
  * coherence config requires `:router` field to be set (MyProject.Web.Router)

## 0.3.1  (2016-11-26)

See these `0.3.0` to `0.3.1` [upgrade instructions](https://gist.github.com/smpallen99/2241d3365aedf0ca3eb884f16def0bb1) to bring your existing app up to speed.

* Enhancements
  * Remove unused params from registration path

lib/coherence.ex
defmodule Coherence do
  @moduledoc """
  Coherence is a full featured, configurable authentication and user
  management system for Phoenix, providing a number of optional features
  configured with a installation mix task:

  * Database Authenticatable: handles hashing and storing an encrypted password in the database.
  * Invitable: sends invites to new users with a sign-up link, allowing the user to create their account with their own password.
  * Registerable: allows anonymous users to register a users email address and password.
  * Confirmable: new accounts require clicking a link in a confirmation email.
  * Recoverable: provides a link to generate a password reset link with token expiry.
  * Trackable: saves login statics like login counts, timestamps, and IP address for each user.
  * Lockable: locks an account when a specified number of failed sign-in attempts has been exceeded.
  * Unlockable With Token: provides a link to send yourself an unlock email.
  * Rememberable: provides persistent login with 'Remember me?' check box on login page.

  See the [README](readme.html) file for an overview, installation, and
  setup instructions.

  ### Authenticatable

  Handles hashing and storing an encrypted password in the database.

  Provides `/sessions/new` and `/sessions/delete` routes for logging in and out with
  the appropriate templates and view.

  The following columns are added the `<timestamp>_add_coherence_to_user.exs` migration:

  * :password_hash, :string - the encrypted password

  The following options can be customized (default shown):

      config :coherence,
        password_hash_field: :password_hash    # you must recompile Coherence after changing this


  ### Invitable

  Handles sending invites to new users with a sign-up link, allowing the user to create their account with their own password.

  Provides `/invitations/new` and `invitations/edit` routes for creating a new invitation and creating a new account from the invite email.

  These routes can be configured to require login by using the `coherence_routes :protected` macro in your router.exs file.

  Invitation token timeout will be added in the future.

  The following table is created by the generated `<timestamp>_create_coherence_invitable.exs` migration:

      create table(:invitations) do
        add :name, :string
        add :email, :string
        add :token, :string
      end

  ### Registerable

  Allows anonymous users to register a users email address and password.

  Provides `/registrations/new` and `/registrations/create` routes for creating a new registration.

  Adds a `Register New Account` to the log-in page.

  It is recommended that the :confirmable option is used with :registerable to
  ensure a valid email address is captured.

  ### Confirmable

  Requires a new account be conformed. During registration, a confirmation token is generated and sent to the registering email. This link must be clicked before the user can sign-in.

  Provides `edit` action for the `/confirmations` route.

  The following options can be customized (default shown):

      config :coherence,
        confirmation_token_expire_days: 5,
        allow_unconfirmed_access_for: 0       # number of days. 0 to disable

  ### Recoverable

  Allows users to reset their password using an expiring token send by email.

  Provides `new`, `create`, `edit`, `update` actions for the `/passwords` route.

  Adds a "Forgot your password?" link to the log-in form. When clicked, the user provides their email address and if found, sends a reset password instructions email with a reset link.

  The following options can be customized (default shown):

      config :coherence,
        reset_token_expire_days: 2

  ### Trackable

  Saves login statics like login counts, timestamps, and IP address for each user.

  Adds the following database field to your User model with the generated migration:

      add :sign_in_count, :integer, default: 0  # how many times the user has logged in
      add :current_sign_in_at, :datetime        # the current login timestamp
      add :last_sign_in_at, :datetime           # the timestamp of the previous login
      add :current_sign_in_ip, :string          # the current login IP adddress
      add :last_sign_in_ip, :string             # the IP address of the previous login

  ### Lockable

  Locks an account when a specified number of failed sign-in attempts has been exceeded.

  The following defaults can be changed with the following config entries:

  The following options can be customized (default shown):

      config :coherence,
        unlock_timeout_minutes: 5,
        max_failed_login_attempts: 5

  Adds the following database field to your User model with the generated migration:

      add :failed_attempts, :integer, default: 0
      add :unlock_token, :string
      add :locked_at, :datetime

  ### Unlockable with Token

  Provides a link to send yourself an unlock email. When the user clicks the link, the user is presented a form to enter their email address and password. If the token has not expired and the email and password are valid, a unlock email is sent to the user's email address with an expiring token.

  The following options can be customized (default shown):

      config :coherence,
        unlock_token_expire_minutes: 5

  ### Remember Me

  The `rememberable` option provides persistent login when the 'Remember Me?' box is checked during login.

  With this feature, you will automatically be logged in from the same browser when your current login session dies using a configurable expiring persistent cookie.

  For security, both a token and series number stored in the cookie on initial login. Each new creates a new token, but preserves the series number, providing protection against fraud. As well, both the token and series numbers are hashed before saving them to the database, providing protection if the database is compromised.

  The following options can be customized (default shown):

      config :coherence,
        rememberable_cookie_expire_hours: 2*24,
        login_cookie: "coherence_login"

  The following table is created by the generated `<timestamp>_create_coherence_rememberable.exs` migration:

      create table(:rememberables) do
        add :series_hash, :string
        add :token_hash, :string
        add :token_created_at, :datetime
        add :user_id, references(:users, on_delete: :delete_all)

        timestamps
      end
      create index(:rememberables, [:user_id])
      create index(:rememberables, [:series_hash])
      create index(:rememberables, [:token_hash])
      create unique_index(:rememberables, [:user_id, :series_hash, :token_hash])

  The `--rememberable` install option is not provided in any of the installer group options. You must provide the `--rememberable` option to install the migration and its support.

  ## Mix Tasks

  Backwords compatibility note: For versions of Phoenix previous to 1.3 use `mix coherence.<command>`. The new `mix coh.<command>` better understands the newer Phoenix patterns.

  ### Installer

  The following examples illustrate various configuration scenarios for the install mix task:

      # Install with only the `authenticatable` option
      $ mix coh.install

      # Install all the options except `confirmable` and `invitable`
      $ mix coh.install --full

      # Install all the options except `invitable`
      $ mix coh.install --full-confirmable

      # Install all the options except `confirmable`
      $ mix coh.install --full-invitable

      # Install the `full` options except `lockable` and `trackable`
      $ mix coh.install --full --no-lockable --no-trackable

  And some reinstall examples:

      # Reinstall with defaults (--silent --no-migrations --no-config --confirm-once)
      $ mix coh.install --reinstall

      # Confirm to overwrite files, show instructions, and generate migrations
      $ mix coh.install --reinstall --no-confirm-once --with-migrations

  Run `$ mix help coh.install` for more information.

  ### Clean

  The following examples illustrate how to remove the files created by the installer:

      # Clean all the installed files
      $ mix coh.clean --all

      # Clean only the installed view and template files
      $ mix coh.clean --views --templates

      # Clean all but the models
      $ mix coh.clean --all --no-models

      # Prompt once to confirm the removal
      $ mix coh.clean --all --confirm-once

  After installation, if you later want to remove one more options, here are a couple examples:

      # Clean one option
      $ mix coh.clean --options=recoverable

      # Clean several options without confirmation
      $ mix coh.clicked --no-confirm --options="recoverable unlockable-with-token"

      # Test the uninstaller without removing files
      $ mix coh.clicked --dry-run --options="recoverable unlockable-with-token"

  Run `$ mix help coh.install` or `$ mix help coherence.install` for more information.
  """
  use Application

  alias Coherence.{Config, Messages}

  @doc false
  def start(_type, _args) do
    Coherence.Supervisor.start_link()
  end

  @doc """
  Get the currently logged in user data.
  """
  def current_user(conn), do: conn.assigns[Config.assigns_key()]

  @doc """
  Updates the user login data in the current sessions.

  Other sessions belonging to the same user won't be updated.
  Requires access to the `conn`, which means it can't be called outside of the context of a conn.
  To update all session belonging to the user see `update_user_login/1`.
  """
  def update_user_login(conn, user) do
    apply(
      Config.auth_module(),
      Config.update_login(),
      [conn, user, [id_key: Config.schema_key()]]
    )
  end

  @doc """
  Updates the user login data in all the sessions belonging to the user.

  All sessions belonging to the same user will be updated.
  Doesn't need access to the `conn`, which means it can be called anywhere.
  To update only the current session see `update_user_login/2`
  """
  def update_user_logins(user) do
    # Handle a user's DBStore
    Coherence.CredentialStore.Server.update_user_logins(user)
  end

  @doc """
  Get the currently logged in user name.
  """
  def current_user_name(conn, field \\ :name) do
    (current_user(conn) || %{}) |> Map.get(field)
  end

  @doc """
  Get the currently assigned user_token
  """
  def user_token(conn), do: conn.assigns[Config.token_assigns_key()]

  @doc """
  Verify a user token for channel authentication.
  """
  def verify_user_token(socket, token, assign_fun) do
    result =
      case Config.verify_user_token() do
        fun when is_function(fun) ->
          fun.(socket, token)

        {mod, fun, args} ->
          apply(mod, fun, args)

        error ->
          {:error,
           Messages.backend().verify_user_token(
             user_token: Config.verify_user_token(),
             error: error
           )}
      end

    case result do
      {:ok, user_id} -> {:ok, assign_fun.(socket, :user_id, user_id)}
      error -> error
    end
  end

  @doc """
  Check if user is logged in.
  """
  def logged_in?(conn), do: !!current_user(conn)
end
